sklearn/utils/validation.py

```diff
@@ -562,6 +568,15 @@ def check_array(array, accept_sparse=False, dtype="numeric", order=None,
     if copy and np.may_share_memory(array, array_orig):
         array = np.array(array, dtype=dtype, order=order)
 
+    if warn_on_dtype and dtypes_orig is not None and {array.dtype} != \
+            set(dtypes_orig):
+        # if there was at the beginning some other types than the final one
+        # (for instance in a DataFrame that can contain several dtypes) then
+        # some data must have been converted
+        msg = ("Data with input dtype %s were all converted to %s%s."
+               % (', '.join(map(str, set(dtypes_orig))), array.dtype, context))
+        warnings.warn(msg, DataConversionWarning, stacklevel=2)
```

Comment:
The other DataConversionWarning does not use `stacklevel`. Generally, I find stacklevel unhelpful. (The best way to get the stack info correct is to run again with an `'error'` filter.)

Reply:
I personally find correct stacklevels very useful (typically when not at the interactive interpreter), however, since this is in `check_array` which is then still used in other sklearn methods (and not directly by the user), I agree this is less useful (or at least should be more than 2)